### PR TITLE
TimeLock + Async Initialization, Part 1

### LINKED ISSUE #2500: Async Initialization Doesn't Work With TimeLock + Cassandra KVS
This happens because the automatic migrator needs access to the KVS, yet it attempts to perform the migration immediately when the TM is created rather than scheduling it to be done async.

```
Exception in thread "main" com.palantir.exception.NotInitializedException: The CassandraKeyValueService is not initialized yet
	at com.palantir.async.initializer.AsyncInitializer.checkInitialized(AsyncInitializer.java:132)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl$InitializingWrapper.delegate(CassandraKeyValueServiceImpl.java:155)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueServiceImpl$InitializingWrapper.delegate(CassandraKeyValueServiceImpl.java:151)
	at com.palantir.atlasdb.keyvalue.cassandra.AutoDelegate_CassandraKeyValueService.createTable(AutoDelegate_CassandraKeyValueService.java:143)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraTimestampBackupRunner.ensureTimestampTableExists(CassandraTimestampBackupRunner.java:55)
	at com.palantir.atlasdb.keyvalue.cassandra.CassandraTimestampStoreInvalidator.backupAndInvalidate(CassandraTimestampStoreInvalidator.java:42)
	at com.palantir.atlasdb.factory.startup.TimeLockMigrator.migrate(TimeLockMigrator.java:64)
	at com.palantir.atlasdb.factory.TransactionManagers.createRawServicesFromTimeLock(TransactionManagers.java:488)
	at com.palantir.atlasdb.factory.TransactionManagers.createRawInstrumentedServices(TransactionManagers.java:472)
	at com.palantir.atlasdb.factory.TransactionManagers.createLockAndTimestampServices(TransactionManagers.java:429)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:226)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:189)
	at com.palantir.atlasdb.factory.TransactionManagers.create(TransactionManagers.java:175)
[redacted]
	at io.dropwizard.cli.EnvironmentCommand.run(EnvironmentCommand.java:40)
	at io.dropwizard.cli.ConfiguredCommand.run(ConfiguredCommand.java:77)
	at io.dropwizard.cli.Cli.run(Cli.java:70)
	at io.dropwizard.Application.run(Application.java:80)
[redacted]
```

### PR DESCRIPTION
**Pull request by jeremyk-91**
**Goals (and why)**:
- Previously, for Cassandra we would attempt to immediately migrate the timestamp bound from Cassandra to TimeLock on startup, which would fail if either of them was unavailable. For DBKVS or other key-value services, we would attempt to ping TimeLock on startup, which would fail if TimeLock was available (though the KVS could be unavailable).
- Fix #2500
- Drive by fix of a thread leak in the async initializer (it doesn't shut down the executor service after successfully initializing the client)

**Implementation Description (bullets)**:
- Make TimeLockMigrator an AsyncInitializer
- Change SerializableTransactionManager to take a Supplier<Boolean> instead of an AsyncInitializer for its TransactionManagersInitializer check
- Require both TLM and TMI to be happy before that prerequisite passes.

**Concerns (what feedback would you like?)**:
- Is the TransactionManagers part too hacky?
- Are there any pitfalls I've missed out on?

**Where should we start reviewing?**: TimeLockMigrator

**Priority (whenever / two weeks / yesterday)**: next week, this would probably block GA of Async Initialization

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2520)

### CODE REVIEW

File: atlasdb-config/src/test/java/com/palantir/atlasdb/factory/startup/TimeLockMigratorTest.java [hunk 1]
```diff
@@ -108,4 +119,60 @@ public void migrationDoesNotProceedIfInvalidationFails() {
...
+        verify(invalidator, times(1)).backupAndInvalidate();
+        wireMockRule.verify(postRequestedFor(urlEqualTo(TEST_ENDPOINT)));
+    }
+
+    @Test
+    public void asyncMigrationProceedsIfInvalidatorInitiallyUnavailable() throws InterruptedException {
+        when(invalidator.backupAndInvalidate())
+                .thenAnswer(new Answer<Long>() {
+                    private AtomicBoolean shouldFail = new AtomicBoolean(true);
+                    @Override
```
Comment (gsheasby) [inline atlasdb-config/src/test/java/com/palantir/atlasdb/factory/startup/TimeLockMigratorTest.java]: nit: newline preferred here (possibly required by baseline?)

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java [hunk 2]
```diff
@@ -390,6 +391,17 @@ SerializableTransactionManager serializable() {
...
     }
 
+    private static boolean areTransactionManagerInitializationPrerequisitesSatisfied(
+            AsyncInitializer initializer,
+            LockAndTimestampServices lockAndTimestampServices) {
+        return initializer.isInitialized() && timeLockMigrationCompleteIfNeeded(lockAndTimestampServices);
+    }
+
+    @VisibleForTesting
+    static boolean timeLockMigrationCompleteIfNeeded(LockAndTimestampServices lockAndTimestampServices) {
```
Comment (gsheasby) [inline atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Idea to consider: you could avoid the boolean supplier and instead have a `Collection<AsyncInitializer>`, if you passed in `initializer` and `lockAndTimestampServices.migrator()`.

File: atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java [hunk 3]
```diff
@@ -71,6 +71,7 @@ private void tryInitializationLoop() {
                     SafeArg.of("className", getInitializingClassName()),
                     SafeArg.of("numberOfAttempts", numberOfInitializationAttempts),
                     SafeArg.of("initializationDuration", System.currentTimeMillis() - initializationStartTime));
+            singleThreadedExecutor.shutdown();
```
Comment (gmaretic) [inline atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java]: 🎉

File: atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java (hunk 2 above)
Comment (gmaretic) [reply atlasdb-config/src/main/java/com/palantir/atlasdb/factory/TransactionManagers.java]: Alternatively, you could have all the initialization prerequisite logic in TransactionManagers. Then `isInitialized()` of the `InitializeCheckingWrapper` can simply be `initializationPrerequisite.get()`. Passing in the migrator just to check its status is a bit strange, but both of these seem better than the "some suppliers, some asyncinitializers" approach IMO.

That said, do not feel strongly about this, so I'm also fine with leaving it as it is.

**Comment by gsheasby** [comment]
This is still blocked on checkstyle :(

**Comment by gmaretic** [comment]
@gsheasby Fixed the checkstyle errors. Since this is high priority, should we merge it in as-is, and leave an issue about possibly refactoring?

**Comment by gsheasby** [comment]
@gmaretic you need to ignore the "final class" warnings rather than fixing them, as they don't play well with Mockito :-(

**Pull request merged**